src/Clash/Cores/Rgmii.hs

```diff
@@ -0,0 +1,216 @@
+  Signal dom (Bool, Maybe (BitVector 8))
+rgmiiReceiver RgmiiRxChannel{..} rxdelay iddr = bundle (ethRxErr, ethRxData)
+ where
+  ethRxCtl :: Signal dom (Bool, Bool)
+  ethRxCtl = iddr rgmiiRxClk resetGen (rxdelay (bitToBool <$> rgmiiRxCtl))
+
+  -- The RXCTL signal at the falling edge is the XOR of RXDV and RXERR
+  -- meaning that RXERR is the XOR of it and RXDV.
+  -- See RGMII interface documentation.
+  ethRxDv, ethRxErr :: Signal dom Bool
+  (ethRxDv, ethRxErr) = unbundle ((\(dv, err) -> (dv, dv `xor` err)) <$> ethRxCtl)
+
+  -- LSB first! See RGMII interface documentation.
+  ethRxData1, ethRxData2 :: Signal dom (BitVector 4)
+  (ethRxData2, ethRxData1) = unbundle $ iddr rgmiiRxClk resetGen (rxdelay rgmiiRxData)
+
+  ethRxData :: Signal dom (Maybe (BitVector 8))
+  ethRxData =
+    (\(dv, dat) -> if dv then Just dat else Nothing)
+      <$> bundle (ethRxDv, liftA2 (++#) ethRxData1 ethRxData2)
+
+-- | RGMII sender. Does not consider transmission error.
+rgmiiSender ::
+  forall dom domDDR.
+  (DomainPeriod dom ~ (*) 2 (DomainPeriod domDDR)) =>
+  Clock dom ->
+  Reset dom ->
+  -- | TX delay function
+  (forall a. Signal domDDR a -> Signal domDDR a) ->
+  -- | oddr function
+  ( forall a.
+    (NFDataX a, BitPack a) =>
+    Clock dom ->
+    Reset dom ->
+    Signal dom a ->
+    Signal dom a ->
+    Signal domDDR a
+  ) ->
+  -- | Maybe the byte we have to send
+  Signal dom (Maybe (BitVector 8)) ->
+  -- | Error signal indicating whether the current packet is corrupt
+  Signal dom Bool ->
+  -- | TX channel to the RGMII PHY
+  RgmiiTxChannel domDDR
+rgmiiSender txClk rst txdelay oddr input err = channel
+ where
+  txEn, txErr :: Signal dom Bit
+  txEn = boolToBit . isJust <$> input
+  txErr = fmap boolToBit err
+
+  ethTxData1, ethTxData2 :: Signal dom (BitVector 4)
+  (ethTxData1, ethTxData2) = unbundle $ maybe (undefined, undefined) split <$> input
+
+  -- The TXCTL signal at the falling edge is the XOR of TXEN and TXERR
+  -- meaning that TXERR is the XOR of it and TXEN.
+  -- See RGMII interface documentation.
+  txCtl :: Signal domDDR Bit
+  txCtl = oddr txClk rst txEn (liftA2 xor txEn txErr)
+
+  -- LSB first! See RGMII interface documentation.
+  txData :: Signal domDDR (BitVector 4)
+  txData = oddr txClk rst ethTxData2 ethTxData1
+
+  channel =
+    RgmiiTxChannel
+      { rgmiiTxClk = txdelay (oddr txClk rst (pure 1) (pure 0))
+      , rgmiiTxCtl = txdelay txCtl
+      , rgmiiTxData = txdelay txData
+      }
+
+{- |
+Circuit that adapts an `RgmiiRxChannel` to a `PacketStream`. Forwards data
+from the RGMII receiver with one clock cycle latency so that we can properly
+mark the last transfer of a packet: if we received valid data from the RGMII
+receiver in the last clock cycle and the data in the current clock cycle is
+invalid, we set `_last`. If the RGMII receiver gives an error, we set `_abort`.
+
+__UNSAFE__: ignores backpressure, because the RGMII PHY is unable to handle that.
+-}
+unsafeRgmiiRxC ::
```

Comment:
Empty UDP packets are not a problem for Ethernet because of padding. But, we should of course not rely on padding as maybe some other protocols have bigger headers or don't do padding.

I have experimented with changing the type of `_last` to `Maybe (Index (dataWidth + 1))`, so that its meaning becomes "the amount of valid bytes in the transaction". It was surprisingly easy to implement for all the generic components in `clash-protocols`: it only took me 2 hours to adjust everything. So, it's looking very promising. The next step is to also adapt our Ethernet components, and test whether it works in hardware.